v0/destinations/google_ads_enhanced_conversion/transform.js

```diff
@@ -0,0 +1,122 @@
+const {
+  getSuccessRespEvents,
+  getErrorRespEvents,
+  CustomError,
+  constructPayload,
+  defaultRequestConfig,
+  getValueFromMessage
+} = require("../../util");
+
+const { trackMapping, BASE_ENDPOINT } = require("./config");
+
+const responseBuilder = async (metadata, message, { Config }, payload) => {
+  const response = defaultRequestConfig();
+  const { event } = message;
+  let flag = 0;
+  const { listOfConversions } = Config;
+  for (let i = 0; i < listOfConversions.length; i += 1) {
+    if (listOfConversions[i].conversions === event) {
+      flag = 1;
+      break;
+    }
+  }
+  if (event === undefined || event === "" || flag === 0) {
+    throw new CustomError(
+      `[Google_ads_enhanced_marketing]:: Conversion named ${event} is not exist in rudderstack dashboard`,
+      400
+    );
+  }
+
+  response.endpoint = `${BASE_ENDPOINT}/${Config.customerId}:uploadConversionAdjustments`;
+  response.body.JSON = payload;
+  const accessToken = metadata.secret.access_token;
+  response.headers = {
+    Authorization: `Bearer ${accessToken}`,
+    "Content-Type": "application/json",
+    "developer-token": getValueFromMessage(metadata, "secret.developer_token")
+  };
+  response.params = { event, customerId: Config.customerId };
+  if (Config.subAccount)
+    if (Config.loginCustomerId)
+      response.headers["login-customer-id"] = Config.loginCustomerId;
+    else
+      throw new CustomError(
+        `[Google_ads_enhanced_marketing]:: loginCustomerId is required as subAccount is true.`,
+        400
+      );
+  return response;
+};
+
+const processTrackEvent = async (metadata, message, destination) => {
+  const payload = constructPayload(message, trackMapping);
+  if (!payload.conversionAdjustments[0].userIdentifiers) {
+    throw new CustomError(
+      `[Google_ads_enhanced_marketing]:: Any of email, phone, firstName, lastName, city, street, countryCode, postalCode or streetAddress is required in traits.`,
+      400
+    );
+  }
+  if (typeof payload.partialFailure !== "boolean") {
+    throw new CustomError(
+      `[Google_ads_enhanced_marketing]:: partialFailure should be a boolean.`,
+      400
+    );
+  }
+  payload.conversionAdjustments[0].adjustmentType = "ENHANCEMENT";
+  // Removing the null values from userIdentifier
+  const arr = payload.conversionAdjustments[0].userIdentifiers;
```

Comment:
As we are hashing the user identifiers would it be possible to hash null values? If not this check is unnecessary. Also if there is null value can we consider hashing empty string as user identifier?

Reply:
This check is basically creating the userIdentifier object. If you check in trackConfig.json, there I am mapping every userIdentifiers position-wise. So if some of the userIdentifier is not exist then I am putting a null value. After that, I am removing all null values from the array.